src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy

```diff
@@ -0,0 +1,251 @@
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.protobuf.gradle
+
+import com.google.common.collect.Maps
+import org.gradle.api.*
+import org.gradle.api.file.FileCollection
+import org.gradle.api.file.SourceDirectorySet
+import org.gradle.api.internal.file.DefaultSourceDirectorySet
+import org.gradle.api.internal.file.FileResolver
+import org.gradle.api.plugins.ExtensionContainer
+import org.gradle.api.tasks.TaskContainer
+import org.gradle.api.tasks.util.PatternSet
+import org.gradle.internal.reflect.Instantiator
+import org.gradle.internal.service.ServiceRegistry
+import org.gradle.language.base.LanguageSourceSet
+import org.gradle.language.base.internal.SourceTransformTaskConfig
+import org.gradle.language.base.internal.registry.LanguageTransform
+import org.gradle.language.base.internal.registry.LanguageTransformContainer
+import org.gradle.language.base.plugins.ComponentModelBasePlugin
+import org.gradle.language.base.sources.BaseLanguageSourceSet
+import org.gradle.language.cpp.CppSourceSet
+import org.gradle.language.cpp.internal.DefaultCppSourceSet
+import org.gradle.language.cpp.tasks.CppCompile
+import org.gradle.model.*
+import org.gradle.nativeplatform.NativeBinarySpec
+import org.gradle.nativeplatform.NativeComponentSpec
+import org.gradle.nativeplatform.NativeDependencySet
+import org.gradle.platform.base.BinarySpec
+import org.gradle.platform.base.LanguageType
+import org.gradle.platform.base.LanguageTypeBuilder
+import org.gradle.platform.base.TransformationFileType
+import org.gradle.util.CollectionUtils
+
+import javax.inject.Inject
+import java.util.concurrent.Callable
+
+class ProtobufModelPlugin implements Plugin<Project> {
+  private final FileResolver fileResolver
+
+  @Inject
+  public ProtobufModelPlugin(FileResolver fileResolver) {
+    this.fileResolver = fileResolver;
+  }
+
+  @Override
+  void apply(Project project) {
+    def gv = project.gradle.gradleVersion =~ "(\\d*)\\.(\\d*).*"
+    if (!gv || !gv.matches() || gv.group(1).toInteger() < 2 || gv.group(2).toInteger() < 7) {
+      println("You are using Gradle ${project.gradle.gradleVersion}: This version of the protobuf" +
+          " plugin for c++ requires minimum Gradle version 2.7")
+    }
+
+    // Provides the osdetector extension
+    project.pluginManager.apply('osdetector')
+
+    project.extensions.create("protobuf", ProtobufConfigurator, project, fileResolver);
+
+    project.pluginManager.apply(ComponentModelBasePlugin)
+  }
+
+  static class Rules extends RuleSource {
```

Comment:
That sounds about right. Just some little note to make sure nothing was forgotten.

 * `finishConfiguringTheGenerateProtoTask()` will be called after `configureGeneratedSourceSets()` because of the mutation dependencies `@Finalize` vs `@Mutate`.
 * Before the rules `registerLanguageTransformation()` and `configureGeneratedSourceSets()`, the rule `resolveProtobufTools()` will be called as it take the `ProtobufConfigurator` as it's subject.
 * The `@LanguageType` annotation is a special one for doing enhancement of the core model. This one in particular register a new language type so we can create new language like:
```
sources {
  anotherSourceSet(ProtobufSourceSet) {...}
}
```
 * As for the rules that take `ModelMap<NativeComponentSpec>` as the subject, they will not necessarily configure the `ModelMap` completely as that type is model aware and will configure the item underneath it only when they are required. With the statement `components.all`, we are only stating that if anything in the `ModelMap<NativeComponentSpec>` is needed, apply the following action to it. If you have multiple component but you are only building a single one, the rules taking `ModelMap<NativeComponentSpec>` as the subject will be called but the action will only be called once as only one component is needed.

To have some additional insight on the rules and model elements, you can use the ./gradlew model which will display the various model elements and what rule is modifying them (and which plugin they are coming from). Very useful for seeing the bigger picture.

I hope this help a bit more, don't hesitate if you have more question.